daemon/mgr/container.go

```diff
@@ -1215,118 +1217,197 @@ func (mgr *ContainerManager) updateContainerResources(c *Container, resources ty
 	return nil
 }
 
-// Upgrade upgrades a container with new image and args.
-func (mgr *ContainerManager) Upgrade(ctx context.Context, name string, config *types.ContainerUpgradeConfig) error {
-	c, err := mgr.container(name)
+func (mgr *ContainerManager) getImageRef(ctx context.Context, image string) (digest.Digest, string, error) {
+	imageID, _, primaryRef, err := mgr.ImageMgr.CheckReference(ctx, image)
+	if err != nil {
+		return "", "", err
+	}
+
+	return imageID, primaryRef.String(), nil
+}
+
+func (mgr *ContainerManager) getOCIImageConfig(ctx context.Context, image string) (ocispec.ImageConfig, error) {
+	img, err := mgr.Client.GetImage(ctx, image)
+	if err != nil {
+		return ocispec.ImageConfig{}, err
+	}
+	ociImage, err := containerdImageToOciImage(ctx, img)
+	if err != nil {
+		return ocispec.ImageConfig{}, err
+	}
+	return ociImage.Config, nil
+}
+
+func (mgr *ContainerManager) prepareContainerEntrypointForUpgrade(ctx context.Context, c *Container, config *types.ContainerUpgradeConfig) error {
```

Comment:
For all the related Upgrade functions. could we move all of them to a saperated file `container_upgrade.go` ? 
Just a piece of advice.